File: frontend/src/api/stats.ts
import { http } from './client'

export type ChartGranularity = 'month' | 'year'
export type ChartMetric = 'quantity' | 'amount'

export interface ShipperProductChart {
  granularity: ChartGranularity
  metric: ChartMetric
  categories: string[]
  series: { name: string; data: number[] }[]
}

export interface TopProductItem {
  product_name: string
  count: number
  amount: string | number
}

export interface ShipperActivity {
  shipper_id: number
  shipper_name: string
  order_count: number
  delivered_count: number
  total_spent: string | number
  avg_order_value: string | number
  orders_per_week: string | number
  top_products: TopProductItem[]
}

export interface DrilldownOrderItem {
  id: number
  order_no: string
  order_date: string
  status: string
  shipper_name: string | null
  product_name: string
  quantity: number
  line_total: string | number
}

export interface DriverPerformanceRow {
  driver_id: number
  driver_name: string
  completed_count: number
  on_time_rate: number | null
  avg_delivery_seconds: number | null
  photo_upload_rate: number
}

export interface DriverPerformance {
  period_label: string
  drivers: DriverPerformanceRow[]
}

export interface ExceptionOrderItem {
  id: number
  order_no: string
  order_date: string
  status: string
  shipper_name: string | null
  driver_name: string | null
  exception_reason: string
  exception_resolution: string
  expected_deliver_before: string | null
  delivered_at: string | null
}

export interface StatsExportBody {
  date_from: string
  date_to: string
  include_shipper_chart?: boolean
  include_driver_perf?: boolean
  include_exceptions?: boolean
  chart_metric?: ChartMetric
  chart_granularity?: ChartGranularity
}

export async function fetchShipperProductChart(params: {
  date_from: string
  date_to: string
  granularity: ChartGranularity
  metric: ChartMetric
}) {
  const { data } = await http.get<ShipperProductChart>('/stats/shipper-product-chart', { params })
  return data
}

export async function fetchShipperActivity(params: {
  shipper_id: number
  date_from: string
  date_to: string
}) {
  const { data } = await http.get<ShipperActivity>('/stats/shipper-activity', { params })
  return data
}

export async function fetchProductDrilldown(params: {
  product_name: string
  date_from: string
  date_to: string
}) {
  const { data } = await http.get<DrilldownOrderItem[]>('/stats/product-drilldown', { params })
  return data
}

export async function fetchDriverPerformance(params: { date_from: string; date_to: string }) {
  const { data } = await http.get<DriverPerformance>('/stats/driver-performance', { params })
  return data
}

export async function fetchExceptionOrders(params: { date_from: string; date_to: string }) {
  const { data } = await http.get<ExceptionOrderItem[]>('/stats/exception-orders', { params })
  return data
}

export async function downloadStatsExport(body: StatsExportBody) {
  const res = await http.post('/stats/export', body, { responseType: 'blob' })
  const blob = res.data as Blob
  const cd = res.headers['content-disposition'] as string | undefined
  let name = `stats-${body.date_from}-${body.date_to}.xlsx`
  if (cd) {
    const m = /filename="?([^";]+)"?/i.exec(cd)
    if (m?.[1]) name = m[1]
  }
  const url = URL.createObjectURL(blob)
  const a = document.createElement('a')
  a.href = url
  a.download = name
  a.click()
  URL.revokeObjectURL(url)
}

File: backend/app/schemas/stats.py
from datetime import date, datetime
from decimal import Decimal
from typing import Any, Literal

from pydantic import BaseModel, Field


class ShipperProductChartOut(BaseModel):
    granularity: Literal["month", "year"] = "month"
    metric: Literal["quantity", "amount"] = "quantity"
    categories: list[str] = Field(..., description="时间轴标签")
    series: list[dict[str, Any]] = Field(..., description="[{name: 商品名, data: [数值按 categories 对齐]}]")


class TopProductItem(BaseModel):
    product_name: str
    count: int
    amount: Decimal


class ShipperActivityOut(BaseModel):
    shipper_id: int
    shipper_name: str
    order_count: int
    delivered_count: int
    total_spent: Decimal
    avg_order_value: Decimal
    orders_per_week: Decimal
    top_products: list[TopProductItem]


class DrilldownOrderItem(BaseModel):
    id: int
    order_no: str
    order_date: date
    status: str
    shipper_name: str | None
    product_name: str
    quantity: int
    line_total: Decimal


class DriverPerformanceRow(BaseModel):
    driver_id: int
    driver_name: str
    completed_count: int
    on_time_rate: float | None = Field(None, description="0~1，无约定/无样本时为 null")
    avg_delivery_seconds: float | None = None
    photo_upload_rate: float = Field(..., description="0~1")
    billing_mode: str | None = Field(None, description="PIECE 计件 / SALARY 工资制")
    freight_owed: str | None = Field(None, description="计件司机待结运费(已结以外)字符串，工资制为 None")


class DriverPerformanceOut(BaseModel):
    period_label: str
    drivers: list[DriverPerformanceRow]


class ShipperPerformanceRow(BaseModel):
    shipper_id: int | None = Field(None, description="无系统账号的临时货主为 null（按 temp_shipper_name 归组）")
    shipper_name: str
    order_count: int = Field(..., description="已送达订单数")
    total_amount: Decimal = Field(..., description="订单明细金额合计（line_total 求和）")


class ShipperPerformanceOut(BaseModel):
    period_label: str
    shippers: list[ShipperPerformanceRow]


class ExceptionOrderItem(BaseModel):
    id: int
    order_no: str
    order_date: date
    status: str
    shipper_name: str | None
    driver_name: str | None
    exception_reason: str
    exception_resolution: str
    expected_deliver_before: datetime | None
    exception_resolved_at: datetime | None = None
    delivered_at: datetime | None


class StatsExportBody(BaseModel):
    date_from: date
    date_to: date
    include_shipper_chart: bool = True
    include_driver_perf: bool = True
    include_exceptions: bool = True
    chart_metric: Literal["quantity", "amount"] = "quantity"
    chart_granularity: Literal["month", "year"] = "month"